README.md
OpenVPN DB Log README
=====================

OpenVPN DB Log is a project to log OpenVPN connect and disconnect events to a
database.

Overview
--------

This project logs OpenVPN connect / disconnect events to a database. The code is
in Perl, so adding support for new backend databases is fairly simple.

The OpenVPN DB Log project is licensed under the GPLv3 license:

* http://opensource.org/licenses/GPL-3.0

Requirements
------------

To use this project meaningfully, you will need:

 * OpenVPN, without which this project won't be of much use
 * Perl installed, and available to execute this project's script
 * The DBI Perl module, plus the relevant DBD:: module for your database
 * The SQL server you are connecting to must be configured for access
 * If this is a new setup, you must load the project schema into the database

Quickstart
----------

For the impatient, a minimal setup includes a connect and disconnect association
with OpenVPN, plus telling it about the database you want to use. This example
uses SQLite, but you're free to use any available backend.

You'll need to create a new database with the schema. For SQLite, you can do:

    sqlite3 /var/db/vpn.sqlite < ./schema/sqlite.sql

Then add the following 2 lines in your OpenVPN config file to provide a basic
logging setup, using the same database:

    client-connect /path/to/openvpn-db-log.pl -b SQLite -d /var/db/vpn.sqlite

    client-disconnect /path/to/openvpn-db-log.pl -b SQLite -d /var/db/vpn.sqlite

Additional features for database backends, recording the server program
instance, and live-client-updates are described in more detail below.

Databases supported
-------------------

Multiple databases are supported by available Perl DBD backends as available on
the local system. So long as you have the DBD driver available and a schema, the
code should work with most standard SQL systems. Schemas are provided in the
`schemas/` directory for the following database systems:

 * MySQL (backend name `mysql`)
 * PostgreSQL (backend name `Pg`)
 * SQLite (backend name `SQLite`)

If there isn't a schema available for your preferred RDBMS, consider creating
and contributing one (more info in docs/Hacking.md.)

Most database backends will require a server, database name, and user
credentials. See the `Database options` help output for the program flags for
each. If your particular database doesn't require one of these, simply omit the
option.

#### Database credentials

  If your database backend requires credentials, you can supply them either with
  the --user / --pass options, or with a --credentials (-c) file. When used, the
  first 2 lines of a credentials file will be taken as the user and password.

Standard features
-----------------

In addition to database connection options, some standard features are
available.

#### Forking: allow user connections if logging fails

  Normally, on-connect processing will report a non-zero status if SQL logging
  fails. This will cause OpenVPN to reject the connection when the script is
  called using the --client-connect OpenVPN directive.

  Where this is undesirable and you want to allow the client connection anyway
  (simply unlogged) you should enable the --fork (-f) option.

  This causes backgrounding of the logging before the SQL connection, returning
  code 0 (success) to OpenVPN so the client connection may continue.

  Note that the script may still exit non-zero if a failure occurs before
  reaching the SQL connect stage. Read about the `Zero` feature to avoid this.

#### Zero Return: always exit code 0 no matter what

  Some implementations may want a 0 exit code in the face of fatal problems,
  such as missing env-vars or bogus command options.

  For this, enable the --zero (-z) flag. Error messages are still printed to
  standard error.

#### Quiet: do not print error text

  To disable printing of error messages, pass the --quiet (-q) flag.

  The exit code is not changed, but can be combined with the --zero flag.

Custom database DSN options
---------------------------

As an advanced/alternative form to provide database values like the database
name, host, port, etc, the --dsn command line option allows you to set any DSN
attribute you would like. This also works for non-standard values if your
specific database backend requires one.

The form of this option is: `--dsn opt=value`

This method can NOT be used to define the backend driver (-b) or the user/pass
(-u/-p) options. This is only valid for the database DSN options.

For example, the following two calls are identical:

    openvpn-db-log -b mysql -d vpn_db -H localhost -t 3306 -u db_user -p db_pass

    openvpn-db-log -b mysql --dsn database=vpn_db --dsn host=localhost \
      --dsn port=3306 -u db_user -p db_pass

Custom database env-vars
------------------------

Another advanced feature used by some DBD drivers is environmental variables. To
set these, the --env option is available, and used like so:

    --env var=value

This will set the env-var `var` to the value `value`. If the value has spaces,
it is necessary to quote the entire stanza, as in: `--env "var=a b c"`.

Notably, the PostgreSQL driver offers PGSYSCONFDIR that is not available as a
DSN option. Also note that this env-var feature can **not** be used to provide
user/pass credentials; you shouldn't do that anyway as this option is exposed to
tools like `ps`.

Processing status files
-----------------------

In order to provide more up-to-date logs and session data, it is possible to
parse an OpenVPN status-file and update the duration and byte-transfer fields
for connected clients. This allows a "partial-update" to occur, even though the
time of disconnect is not yet known.

This is especially useful in environments where long-lived connections are
common. When an OpenVPN server is terminated (or the app/OS crashes) there will
be no disconnect event despite a partial database row for the connection.

#### Status file requirements

  In order to do partial-updates of status files, you must:

  * Enable the `--status-file` OpenVPN directive
  * Set the `--status-version` to version 2 or 3 (v1 is not supported)
  * Enable regular processing of the status file (see below)

#### Sending the status file for log processing

  You will need to set up some regular process by which the program is told to
  parse a status file from OpenVPN; this could be cron, or some other on-demand
  event that suits your needs.

  The --status-file (-S) flag is the file path and the --status-version (-V)
  flag must match the OpenVPN setting by the same name (v3 is the default.)

  The --update-create (-C) flag allows a missing session entry to be created
  instead of failing that entry. This situation could occur if a prior
  client-connect event failed (due to DB error or network issues.) This feature
  enables the session to be added at the time of update.

  As an advanced feature, the -N flag mandates that every client line must be
  successfully matched and added to the database in order for the transaction to
  go through.  This is probably only useful for people processing the real-time
  output of a `status` command from OpenVPN using the STDIN input support; and
  only then when this is desired behavior.

#### Allowable status file age

  The --status-age (-A) flag sets a maximum allowable increase between the
  status file timestamp and the current system clock, in seconds. If an OpenVPN
  server has terminated, the status file is often left as-is; this option
  prevents needless database connections.

#### Increasing status file verbosity

  Additional informational messages for status-file updates can be printed to
  STDERR when the -N option is not used.

  Passing the --status-info (-I) option 1 or more times enables this feature.
  This option takes no arguments. The verbosity levels are:

  1. Invalid --status-file lines show the reason for rejection
  2. Like #1, but also print the full rejected line


Instance support: logging for multiple servers
----------------------------------------------

Some environments have multiple OpenVPN servers and would like to log them to a
single backend. Common examples include running on TCP + UDP, leveraging
multiple CPU cores, or even separate servers being logged centrally.

In order to identify the correct source of connections in a multi-instance
environment, 3 optional instance descriptors are available:

  * `Name`: a textual name or description, up to 64 characters
  * `Protocol`: a textual protocol, up to 10 characters
  * `Port`: a TCP/UDP port in the range of 1-65535

These options default to empty strings and the pseudo-value `0` respectively.

When using this feature, the **unique** combination of these 3 values defines a
separate instance. This means that if you log some connections with name=Blue,
then other connections with name=Blue port=1195, these would be stored
internally with different instance IDs.

#### Changing instance values

  Once an instance association has been added to the database, the unique
  combination appears in the `instances` table. This means updating values is
  possible without much effort, but will require an update to the table entry
  for the instance definition.

  This may eventually be made easier through a utility script. Note that this
  information is stored in the `instances` table alone, and referenced as an ID
  from the `sessions` table.

Partial-update API
------------------

Most uses of the partial-update feature will likely want to use the status
processing feature described above. For advanced needs, a basic API is provided
to do partial-update processing. Limited error checking is done, since the
values are normally supplied from OpenVPN anyway.

To use, declare the following env-vars for the call to openvpn-db-log.pl:

  * `trusted_ip` or `trusted_ip6`
  * `trusted_port`
  * `common_name`
  * `bytes_received`
  * `bytes_sent`
  * `time_unix`
  * `script_type` (must be set to the string: "db-update")
  * `ifconfig_pool_remote_ip` (optional, as OpenVPN may not have issued one)
  * `time_update` (optional, the system time will be used when omitted)

Note that any program CLI options for DB or other program features must still be
supplied as described earlier.

The `--update-create` feature can be used with this API as well (see `Processing
status files` for details.)

Versions and schema stability
-----------------------------

Running release-versions is the safest way to go as this project won't guarantee
schema stability for non-release versions. Official releases will contain
incremental SQL upgrade scripts for any schema changes.

If you're not comfortable adjusting database schema between versions as
necessary, a development branch is probably not for you.



ChangeLog
ChangeLog for OpenVPN DB Log project

If you want an overview of release info/changes, see the ReleaseNotes instead.

* 0.9.0 (2014-08-18)
	* First official release version
	* No upstream support for VPN-issued IPv6 logging
		* This will be included in a future 1.x release when available
	Author: Josh Cepek <josh.cepek@usa.net>



openvpn-db-log.pl
#!/usr/bin/perl

# Log openvpn connections to a DB.
# Handles both connect and disconnect events, keyed by exposed env-vars.

# This file is part of the OpenVPN DB Log project:
# https://github.com/QueuingKoala/openvpn-db-log
#
# Copyright Josh Cepek <josh.cepek@usa.net> 2014
#   with full contributor list in the ChangeLog
#
# Available under the GPLv3 license:
# http://opensource.org/licenses/GPL-3.0

use strict;
use Getopt::Long;
use DBI;
Getopt::Long::Configure ("bundling");
my $prog_version="0.9.0";

sub usage {
	printf <<EOM;
OpenVPN DB Log is a GPLv3 project. See --copyright for details.
Version: $prog_version

OPTIONS:

Database options:
  --backend, -b
      The Perl DBI backend to use. Mandatory.
  --database, --db -d
      The database to connect to.
  --host, -H
      Database host to connect to.
  --port, -t
      Port number to connect to.
  --user, -u
      Database username.
  --password, pass, -p
      Database password.
  --credentials, --cred, -c
      File for database authentication, user/pass on first 2 lines.
  --dsn
      An advanced method to define DB DSN options in the form: opt=value
      See docs for details.
  --env
      Advanced DB feature to set env-vars in the form: var=value
      See docs for details (feature used by some DBI modules.)

Basic options:
  --fork, -f
      After basic option checking, exit code 0 and fork for SQL processing.
  --quiet, -q
      Do not report any errors to STDERR (does not change the exit code.)
  --zero, -z
      Failure exits with code 0, primarily for systems lacking Perl fork().
  --help, --usage, -h
      You're reading it.

Instance options:
  --instance-name, -n (up to 64 chars)
  --instance-proto, -r (up to 10 chars)
  --instance-port, -o
      Optional values to identify a unique OpenVPN instance (see docs.)

Status file processing:
  --status-file, -S
      Path to the status file. Supply an empty string argument for STDIN.
  --status-version, -V
      OpenVPN status format version. Must be 2 or 3, and defaults to 3.
  --status-need-success, -N
      Refuse the update if any client entries fail (see docs.)
  --status-age, -A
      Maximum allowable age in seconds of the status file timestamp.
  --status-info, -I
      Print extra info to STDERR for ignored client lines (see docs.)
  --update-create, -C
      During updates, add the session/instance if there is no matching entry.
EOM
        exit 0;
}

sub license {
	printf <<EOM;
OpenVPN DB Log is available under a GPLv3 license.
Copyright Josh Cepek <josh.cepek\@usa.net> 2014

Full contributor list available in the ChangeLog.
The GPLv3 license can be found in full in the project tree at docs/GPLv3.txt.

This is free software, and you are welcome to redistribute it under certain
conditions. See the full license for details, also available from the FSF or:
http://opensource.org/licenses/GPL-3.0
EOM
	exit 0
}

# Holds data of interest to be logged to the database:
my %data;
# Database vars:
my %db_env;
my %dsn;
my %db = (
	user	=> "",
	pass	=> "",
);
# Common config vars:
my %conf = (
	fork	=> 0,
	quiet	=> 0,
	rc_zero	=> 0,
);
# Instance vars:
my %instance = (
	name	=> '',
	proto	=> '',
	port	=> 0,
);
# Status file vars:
my %status = (
	need_success	=> 0,
	version		=> 3,
	verb		=> 0,
	create		=> 0,
);
GetOptions(
	"fork|f!"		=> \$conf{fork},
	"quiet|q!"		=> \$conf{quiet},
	"zero|z!"		=> \$conf{rc_zero},
	"backend|b=s"		=> \$db{driver},
	"user|u=s"		=> \$db{user},
	"password|pass|p=s"	=> \$db{pass},
	"credentials|creds|c=s"	=> \$db{creds},
	"database|db|d=s"	=> \$dsn{database},
	"host|H=s"		=> \$dsn{host},
	"port|t=i"		=> \$dsn{port},
	"dsn=s"			=> \%dsn,
	"env=s"			=> \%db_env,
	"instance-name|n=s"	=> \$instance{name},
	"instance-proto|r=s"	=> \$instance{proto},
	"instance-port|o=i"	=> \$instance{port},
	"status-file|S:s"	=> \$status{file},
	"status-version|V=i"	=> \$status{version},
	"status-need-success|N"	=> \$status{need_success},
	"status-age|A=i"	=> \$status{age},
	"status-info|I+"	=> \$status{verb},
	"update-create|C"	=> \$status{create},
	"help|usage|h"          => \&usage,
	"copyright|license"	=> \&license,
);

# Verify CLI opts
defined $db{driver}
	or failure("Options error: no backend driver provided");
length($instance{name}) <= 64
	or failure("Options error: instance-name too long (>64)");
length($instance{proto}) <= 10
	or failure("Options error: instance-proto too long (>10)");
$instance{port} >= 0 and $instance{port} <= 65535
	or failure("Options error: instance-port out of range (1-65535)");

read_creds() if defined $db{creds};

# Set any DB-specific env-vars passed:
for my $var (keys %db_env) {
	$ENV{$var} = $db_env{$var};
}

# Status file processing won't continue below
status_proc() if defined $status{file};

# Update %data with required env-vars, keyed by shorter reference names.
# Disconnect/update will add to this hash later if needed
env_opt(src_port	=> 'trusted_port');
env_opt(cn		=> 'common_name');
env_opt(vpn_ip4		=> 'ifconfig_pool_remote_ip', "");

# Append to mandatory env-vars and set sub handler depending on script_type
my $type;
$type = $ENV{script_type}
	or failure("Missing required script_type env-var");
my $handler = \&connect;
if ( $type =~ /^client-disconnect$/ ) {
	$handler = \&disconnect;
	# add some additional vars used during disconnect
	env_opt(time		=> 'time_unix');
	env_opt(duration	=> 'time_duration');
	env_opt(bytes_in	=> 'bytes_received');
	env_opt(bytes_out	=> 'bytes_sent');
}
elsif ( $type =~ /^client-connect$/ ) {
	env_opt(time		=> 'time_unix');
}
elsif ( $type =~ /^db-update$/) {
	$handler = \&update;
	# vars required for updates:
	env_opt(bytes_in	=> 'bytes_received');
	env_opt(bytes_out	=> 'bytes_sent');
}
else {
	failure("Invalid script_type: '$type'");
}

# Need either trusted_ip or trusted_ip6 from env:
$data{src_ip} = $ENV{trusted_ip} || $ENV{trusted_ip6}
	or failure("ERR: missing env-var: trusted_ip");

# When forking, exit success and continue SQL tasks as the child process
db_fork() if ( $conf{fork} );

db_connect();

# Take the right DB update action depending on script type.
# Any database errors escape the eval to be handled below.
eval {
	$handler->();
	$db{dbh}->commit();
};

# Handle any DB transaction errors from the handler sub
db_rollback($@) if ($@);

# Success otherwise
exit 0;

# Exit handler, for message display and return code control
sub failure {
	my ($msg) = @_;
	warn "$msg" if $msg and not $conf{quiet};
	exit 0 if $conf{rc_zero};
	exit 100;
};

# Env-var option helper
# Call as: env_opt( 'opt_name', 'env_var' [, 'default-when-optional']
sub env_opt {
	my ($opt, $env, $default) = @_;
	return if ( $data{$opt} = $ENV{$env} );
	defined $default
		or failure("Error: missing env-var: $env");
	$data{$opt} = $default;
}

# Credentials processing
sub read_creds {
	open(my $fh, "<", $db{creds})
		or failure("Unable to open credentials file");
	($db{user}, $db{pass}) = grep( defined, map(<$fh>, 1..2) );
	defined $db{pass} or failure("Invalid credentials file");
	chomp %db;
}

# Fork handler; closes standard file handles
sub db_fork {
	open(STDIN, "<", "/dev/null");
	open(STDOUT, ">", "/dev/null");
	open(STDERR, ">", "/dev/null");
	fork and exit 0;
}

# Generic DB error handler
sub db_rollback {
	my $msg = shift || "";
	eval { $db{dbh}->rollback; };
	failure($msg);
}

# Connect to the SQL DB
sub db_connect {
	my $driver_dsn = "";
	for my $key (keys %dsn) {
		next unless defined $dsn{$key};
		$driver_dsn .= "$key=$dsn{$key};";
	}
	$driver_dsn =~ s/;$//;
	$db{dbh} = DBI->connect(
		"dbi:$db{driver}:$driver_dsn",
		$db{user},
		$db{pass}, {
			AutoCommit	=> 0,
			PrintError	=> 0,
		}
	);

	# Handle DB connect errors
	defined $db{dbh}
		or failure("DB connection failed: ($DBI::errstr)");
	$db{dbh}->{RaiseError} = 1;
}

# Insert the connect data
sub connect {
	my $iid = get_instance(create => 1);
	$db{dbh}->do(qq{
		INSERT INTO
		session (
			connect_time,
			src_ip,
			src_port,
			vpn_ip4,
			cn,
			instance_id
		)
		VALUES (?, ?, ?, ?, ?, ?)
		},
		undef,
		$data{time},
		$data{src_ip},
		$data{src_port},
		$data{vpn_ip4},
		$data{cn},
		$iid,
	);
}

# Insert the disconnect data
sub disconnect {
	my $sth;
	my $iid = get_instance();
	my $id = match_session_id(iid => $iid);

	# Update session details with disconnect values:
	$data{disconnect_time} = $data{time} + $data{duration};
	$sth = $db{dbh}->do(qq{
		UPDATE
			session
		SET
			disconnect_time = ?,
			duration = ?,
			bytes_in = ?,
			bytes_out = ?
		WHERE
			id = ?
		},
		undef,
		$data{disconnect_time},
		$data{duration},
		$data{bytes_in},
		$data{bytes_out},
		$id,
	);
}

# Update a session
sub update {
	my %f_opt = ( @_ );
	my $iid = $f_opt{iid} || get_instance(create => $status{create});
	my $update_time = $f_opt{time_update} || time();

	my $id = match_session_id( iid => $iid, create => $status{create} );

	# Calculate current duration, and basic sanity check:
	$data{duration} = $update_time - $data{time};
	$data{duration} >= 0 or die "Failed update: time has gone backwards";

	# Prepare update query, unless we have one
	defined $db{sth_update} or $db{sth_update} = $db{dbh}->prepare(qq{
		UPDATE
			session
		SET
			duration = ?,
			bytes_in = ?,
			bytes_out = ?
		WHERE
			id = ?
	});

	# Update session details with supplied values:
	$db{sth_update}->execute(
		$data{duration},
		$data{bytes_in},
		$data{bytes_out},
		$id,
	);
}

# Get ID of an instance.
# When the `create` opt is true, will attempt to create if needed
sub get_instance {
	my %f_opt = (
		create	=> 0,
		@_
	);
	my $sth = $db{dbh}->prepare(qq{
		SELECT	id
		FROM	instance
		WHERE
			name = ?
		  AND	port = ?
		  AND	protocol = ?
		ORDER BY
			id ASC
		LIMIT	1
	});
	$sth->execute(
		$instance{name},
		$instance{port},
		$instance{proto},
	);
	my $id = $sth->fetchrow_array;

	# Try to add instance details if none present
	if ( ! defined $id and ($f_opt{create}) ) {
		$id = add_instance();
	}
	return $id if defined $id;
	die "Failed instance association";
}

sub add_instance {
	$db{dbh}->do(qq{
		INSERT INTO
		instance (
			name,
			port,
			protocol
		)
		values (?, ?, ?)
		},
		undef,
		$instance{name},
		$instance{port},
		$instance{proto},

	);
	return get_instance();
}

# Associate with the connect session using env-vars:
# Must define `iid => <value>`
# Optionally set `create => 1` to call create() for a missing session entry
sub match_session_id {
	my %f_opt = (
		create	=> 0,
		@_
	);
	my @query_opts;

	# vpn_ip4 is optional, so set up the query to support both cases:
	my $vpn_ip_query = "= ?";
	my $sth_name = "sth_session";
	if (not defined $data{vpn_ip4} or length($data{vpn_ip4}) == 0) {
		$vpn_ip_query = "IS NULL";
		$sth_name .= "_null";
	}
	else {
		push @query_opts, $data{vpn_ip4};
	}

	# Prepare session query, unless we have one
	defined $db{$sth_name} or $db{$sth_name} = $db{dbh}->prepare(qq{
		SELECT	id, disconnect_time
		FROM	session
		WHERE
			vpn_ip4 $vpn_ip_query
		  AND	connect_time = ?
		  AND	src_ip = ?
		  AND	src_port = ?
		  AND	cn = ?
		  AND	instance_id = ?
		ORDER BY
			id DESC
		LIMIT	1
	});

	# Then run the query on the client option data
	push @query_opts, (
		$data{time},
		$data{src_ip},
		$data{src_port},
		$data{cn},
		$f_opt{iid},
	);
	$db{$sth_name}->execute(@query_opts);

	# Look for a matching session:
	my ($id, $disconnect) = $db{$sth_name}->fetchrow_array();
	defined $disconnect and die "Cannot update a session that has disconnected";
	# If there was no match, see if it needs to be created:
	if ( not defined $id ) {
		($f_opt{create}) or die "No matching connection entry found";
		&connect;
		$id = match_session_id( @_, create => 0 );
	}

	return $id;
}

# Process a status file
sub status_proc {
	my $input;
	my $delim;
	$delim = "\t" if ($status{version} == 3);
	$delim = "," if ($status{version} == 2);
	defined $delim or failure("Invalid status version: must be 2 or 3");

	if ( length($status{file}) > 0 ) {
		open($input, "<", $status{file})
			or failure("Failed to open '$status{file}' for reading");
	}
	else {
		open($input, "<-")
			or failure("Failed to open STDIN for status reading");
	}

	db_fork() if ( $conf{fork} );

	my @fields;
	my $iid;
	my $time_update;
	my $bad_lines = 0;
	while (<$input>) {
		chomp;
		# pull out time:
		if ( /^TIME$delim.*$delim([0-9]+)$/ ) {
			if ( defined $status{age} ) {
				( time() - $1 <= $status{age} )
					or failure("Status file exceeds aging limit");
			}
			$time_update = $1;
		}
		next unless defined $time_update;

		# Otherwise process client list lines.
		next unless /^CLIENT_LIST($delim.*){8}/;
		%data = ();
		@fields = split /$delim/;
		shift @fields;

		# CN can have a comma, so process records from the right until then.
		for my $key (qw(user time junk bytes_out bytes_in vpn_ip4 remote)) {
			$data{$key} = pop @fields;
		}

		# Remainder is the CN:
		$data{cn} = join('', @fields);

		# pull source IP/port:
		if ( $data{remote} =~ /^(.+):([0-9]+)$/ ) {
			$data{src_ip} = $1;
			$data{src_port} = $2;
		}
		else {
			warn "bad IP/port in input" if ( $status{verb} >= 1 );
			warn " bad line: $_" if ( $status{verb} >= 2 );
			$bad_lines += 1;
			next;
		}

		# Do any delayed DB setup tasks now that we have a real line:
		if ( ! defined $db{dbh} ) {
			eval {
				db_connect();
				$iid = get_instance();
			};
			failure ($@) if ($@);
		}

		# Now perform the update, which uses values assigned to %o:
		eval {
			update( iid => $iid, time_update => $time_update );
		};
		# Error handling:
		# Only do a rollback when 100% success is required:
		db_rollback($@) if ($@) and ($status{need_success});
		# Otherwise just count the failure:
		if ($@) {
			warn "bad input: $@" if ( $status{verb} >= 1 );
			warn " bad line: $_" if ( $status{verb} >= 2 );
			$bad_lines += 1 if ($@);
		}
	}

	# Final DB commit if anything happened:
	eval {
		$db{dbh}->commit() if defined $db{dbh};
	};
	# Error handling:
	db_rollback($@) if ($@);

	exit 0 if $conf{rc_zero};
	$bad_lines = 99 if ($bad_lines > 99);
	exit $bad_lines;
}

